crates/validation/src/lib.rs

```diff
@@ -230,3 +246,15 @@ pub async fn validate(
         errors,
     }
 }
+
+// This pattern lets us cheaply detect if a read schema references the inferred
+// of it's collection. Assuming an otherwise well-formed JSON schema,
+// it can neither false-positive nor false-negative:
+// * It must detect an actual property because a the same pattern within a JSON
+//   string would be quote-escaped.
+// * It must be a schema keyword ($ref cannot be, say, a property) because
+//   "flow://inferred-schema" is not a valid JSON schema and would error at build time.
+const REF_INFERRED_SCHEMA_PATTERN: &str = "\"$ref\":\"flow://inferred-schema\"";
```

Comment:
This is sensitive to whitespace around the `:`, and it took me a while to prove to myself that there aren't any code paths that could result in a `"$ref" : "flow://inferred-schema"`. I _think_ this works because `Loader::load_resource_content` will always re-serialize the schema without whitespace. Assuming that's correct, I think it at least warrants a comment.

I'll suggest going just a little further and factoring this out into `fn uses_*_schema_ref(schema: &str) -> bool` instead of using `schema.contains(REF_*)`, which can carry a comment about schema serialization.

Reply:
You're right, it felt brittle. I've moved this into members of `models::Schema`, using whitespace-invarant regex's under the hood.